README.md
# preact-leaflet

🚧 This library is current under construction. 🚧

Currently Supporting the following Leaflet components:
* [MapContainer](https://leafletjs.com/reference.html#map-example)
* [Marker](https://leafletjs.com/reference.html#marker)
* [TileLayer](https://leafletjs.com/reference.html#tilelayer)

## Example

```jsx
import { h } from 'preact';
import { SvgIcon } from './SvgIcon';
import {
  MapContainer,
  TileLayer,
  Marker,
} from 'preact-leaflet';
import 'leaflet/dist/leaflet.css';
import './styles.css';


export const App = () => {
  return (
    <MapContainer
      // ref={this.mapRef}
      center={[51.505, -0.09]}
      zoom={13}
      options={{
        // Options
        // preferCanvas: false,

        // Control Options
        // attributionControl: true,
        // zoomControl: true,

        // Interaction Options
        // closePopupOnClick: true,
        // zoomSnap: 1,
        // zoomDelta: 1,
        // trackResize: true,
        // boxZoom: true,
        // doubleClickZoom: true,
        // dragging: true,

        // Map State Options
        // crs:
        // center: [51.505, -0.09],
        // zoom: 13,
        // minZoom: // Number
        // maxZoom: // Number
        // layers: [],
        // maxBounds: null,
        // rendered:

        // Animation Options
        // zoomAnimation: true,
        // zoomAnimationThreshold: 4,
        // fadeAnimation: true,
        // markerZoomAnimation: true,
        // transform3DLimit:

        // Panning Inertia Options
        // inertia:
        // inertiaDeceleration:
        // inertiaMaxSpeed:
        // easeLinearity:
        // worldCopyJump:
        // maxBoundsViscosity:

        // Keyboard Navigation Options
        // keyboard:
        // keyboardPanDelta:

        // Mouse wheel options
        // scrollWheelZoom:
        // wheelDebounceTime:
        // wheelPxPerZoomLevel:

        // Touch interaction options
        // tapHold:
        // tapTolerance:
        // touchZoom:
        // bounceAtZoomLimits:
      }}
    >
      <TileLayer
        // to use mapbox
        url="https://api.mapbox.com/styles/v1/{id}/tiles/{z}/{x}/{y}?access_token={accessToken}"
        // to use open street maps
        // url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
        options={{
          // to use mapbox
          attribution: "Map data &copy; <a href='https://www.openstreetmap.org/copyright'>OpenStreetMap</a> contributors, Imagery © <a href='https://www.mapbox.com/'>Mapbox</a>",
          // to use open street maps
          // attribution: '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
          maxZoom: 18,
          id: 'mapbox/streets-v11',
          tileSize: 512,
          zoomOffset: -1,
          // required for mapbox
          accessToken: '',
          // crossOrigin: false,
        }}
      />
      <Marker
        position={[51.5, -0.09]}
        icon={SvgIcon}
        options={{
          title: 'this is my title',
          opacity: 0.5,
        }}
      />
    </MapContainer>
  );
}
```

## API

This project is a thin wrapper around the native [leafletjs](https://leafletjs.com/reference.html) library.


### Options

 - [x] Support for all Map options
 - [x] Support for all Marker options
 - [x] Support for all Tile Layer options

### TODO: Handle Event listeners

 - [ ] Support for all Map events
 - [ ] Support for all Marker events
 - [ ] Support for all Tile Layer events

### TODO:

 - [ ] Add ability to get a ref to Markers and TileLayer


## Contributing
Yes. Do it. All about that.

#### How to contribute
1. Fork the project
2. Create a feature branch (`git checkout -b f/amazingFeature`)
3. Commit your changes (`git commit -m 'added awesome sauce'`)
4. Push to the remote branch (`git push origin f/amazingFeature`)
5. Open a pull request.

#### Contributors: 1

- :monkey_face: Christopher Harold Butler ([ChristopherHButler](https://github.com/ChristopherHButler))

## Acknowledgments

Inspiration for the library was initially taken from the now archived [kontrollanten/preact-leaflet](https://github.com/kontrollanten/preact-leaflet). It has since been completely re-written.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.


 - [x] Support for all Map options
 - [x] Support for all Marker options
 - [x] Support for all Tile Layer options
 - [x] Cleanup Map, Marker and TileLayer (with return from useEffect) 









### 0.1.0 (2019-12-10)

Note: Changes made to verion 0.1.0 and prior were from a different author: [kontrollanten/preact-leaflet](https://github.com/kontrollanten/preact-leaflet). This repo has been forked and modified.
### Features

* **Map:** add support for all leaflet Map options 
* **Map:** add support for event listeners 
* **Map:** add support for passing bounds prop
* **Map:** add support to update center position
* **Map:** add support to update zoom 
* **Map:** set default zoom to 6 
* **Marker:** add support for markers 
* **MarkerCluster:** add marker cluster support 
* **Polyline:** add Polyline support 
* **ZoomControl:** add ZoomControl support 


### Bug Fixes

* make preact X compat 
* **deps:** update dependency preact to v10 
* **eslint:** v18 config 
* add displayName to components 
* **demo:** set accurate title 
* **Layer:** log error instead of throw exception 
* **leaflet:** use named import from leaflet 
* **Map:** prevent render null children 
* **Map:** remove default center value 


demo/App.js
import { h, Component, Fragment } from 'preact';
import 'preact/debug';
import { SvgIcon } from './SvgIcon';
import {
  MapContainer,
  TileLayer,
  Marker,
} from '../dist';
import 'leaflet/dist/leaflet.css';
import './styles.css';
import route from './long-route';

const getRouteConfig = (hash) => {
  switch (hash) {
    case '#cluster':
      return {
        mapCenter: [58.9700, 5.7331],
        markerCluster: [
          [58.9275, 5.6607],
          [58.9275, 5.6607],
          [58.9275, 5.6607],
          [58.9275, 5.6607],
          [58.9275, 5.6607],
          [58.9275, 5.6607],
          [58.9275, 5.6607],
        ],
        markers: [],
        polylines: [],
        zoom: 8,
      };
    case '#polyline':
      return {
        mapCenter: [63.83919, 20.15069],
        markerCluster: false,
        markers: [
          [59.3367, 18.0667],
          [63.8374896962485, 20.163206074534],
        ],
        polylines: [
          route,
        ],
        zoom: 5,
      };
    default:
      return {
        mapCenter: [58.9700, 5.7331],
        markerCluster: false,
        markers: [],
        polylines: [],
        zoom: 10,
      };
  }
};

export default class App extends Component {
  constructor(props) {
    super(props);
    this.state = getRouteConfig(window.location.hash);
  }

  componentDidMount() {
    window.addEventListener('popstate', () => {
      this.setState(getRouteConfig(window.location.hash));
    });
  }

  render() {
    return (
      <Fragment>
        <header>
          <h1>preact-leaflet</h1>
          <p>Better demo coming soon</p>
        </header>
        <MapContainer
          // ref={this.mapRef}
          center={[51.505, -0.09]}
          zoom={13}
          options={{
            // Options
            // preferCanvas: false,

            // Control Options
            // attributionControl: true,
            // zoomControl: true,

            // Interaction Options
            // closePopupOnClick: true,
            // zoomSnap: 1,
            // zoomDelta: 1,
            // trackResize: true,
            // boxZoom: true,
            // doubleClickZoom: true,
            // dragging: true,

            // Map State Options
            // crs:
            // center: [51.505, -0.09],
            // zoom: 13,
            // minZoom: // Number
            // maxZoom: // Number
            // layers: [],
            // maxBounds: null,
            // rendered:

            // Animation Options
            // zoomAnimation: true,
            // zoomAnimationThreshold: 4,
            // fadeAnimation: true,
            // markerZoomAnimation: true,
            // transform3DLimit:

            // Panning Inertia Options
            // inertia:
            // inertiaDeceleration:
            // inertiaMaxSpeed:
            // easeLinearity:
            // worldCopyJump:
            // maxBoundsViscosity:

            // Keyboard Navigation Options
            // keyboard:
            // keyboardPanDelta:

            // Mouse wheel options
            // scrollWheelZoom:
            // wheelDebounceTime:
            // wheelPxPerZoomLevel:

            // Touch interaction options
            // tapHold:
            // tapTolerance:
            // touchZoom:
            // bounceAtZoomLimits:
          }}
        >
          <TileLayer
            // to use mapbox
            url="https://api.mapbox.com/styles/v1/{id}/tiles/{z}/{x}/{y}?access_token={accessToken}"
            // to use open street maps
            // url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"
            options={{
              // to use mapbox
              attribution: "Map data &copy; <a href='https://www.openstreetmap.org/copyright'>OpenStreetMap</a> contributors, Imagery © <a href='https://www.mapbox.com/'>Mapbox</a>",
              // to use open street maps
              // attribution: '&copy; <a href="https://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors',
              maxZoom: 18,
              id: 'mapbox/streets-v11',
              tileSize: 512,
              zoomOffset: -1,
              // required for mapbox
              accessToken: 'pk.eyJ1IjoiY2J1dGxlcjAwIiwiYSI6ImNsMWZodHd2NjAwMDIzZXF1N3I0Zzc4MTcifQ.fi4CpKahbRAP3VlmVC0Q-A',
              // crossOrigin: false,
            }}
          />
          <Marker
            position={[51.5, -0.09]}
            icon={SvgIcon}
            options={{
              title: 'this is my title',
              opacity: 0.5,
            }}
          />
        </MapContainer>
      </Fragment>
    );
  }
}
